spec/spec.md

```diff
@@ -0,0 +1,2789 @@
+must return to the indent level of the parent.
+
+Since conditional statements (`when` and `else`) may be nested, it is
+possible to create a hierarchy of indent levels, each with its own
+number of preceding spaces that must be larger than its parent's and
+consistent among all direct child statements (those that are not
+children of an even deeper conditional statement).
+
+As a concrete guide, a few consequences of these rules are summarized
+below:
+
+-   The `circuit` keyword must not be indented.
+
+-   All `module` keywords must be indented by the same number of spaces.
+
+-   In a module, all port declarations and all statements (that are not
+    children of other statements) must be indented by the same number of
+    spaces.
+
+-   The number of spaces comprising the indent level of a module is
+    specific to each module.
+
+-   The statements comprising a conditional statement's branch must be
+    indented by the same number of spaces.
+
+-   The statements of nested conditional statements establish their own,
+    deeper indent level.
+
+-   Each `when` and each `else` context may have a different number of
+    non-zero spaces in its indent level.
+
+As an example illustrating some of these points, the following is a
+legal FIRRTL circuit:
+
+``` firrtl
+circuit Foo :
+    module Foo :
+      skip
+    module Bar :
+     input a: UInt<1>
+     output b: UInt<1>
+     when a:
+         b <= a
+     else:
+       b <= not(a)
+```
+
+All circuits, modules, ports and statements can optionally be followed
+with the info token `@[fileinfo]` where fileinfo is a string containing
+the source file information from where it was generated. The following
+characters need to be escaped with a leading '`\`': '`\n`' (new line),
+'`\t`' (tab), '`]`' and '`\`' itself.
+
+The following example shows the info tokens included:
+
+``` firrtl
+circuit Top : @[myfile.txt 14:8]
+   module Top : @[myfile.txt 15:2]
+     output out:UInt @[myfile.txt 16:3]
+     input b:UInt<32> @[myfile.txt 17:3]
+     input c:UInt<1> @[myfile.txt 18:3]
+     input d:UInt<16> @[myfile.txt 19:3]
+     wire a:UInt @[myfile.txt 21:8]
+     when c : @[myfile.txt 24:8]
+       a <= b @[myfile.txt 27:16]
+     else :
+       a <= d @[myfile.txt 29:17]
+     out <= add(a,a) @[myfile.txt 34:4]
+```
+
+\pagebreak
+
+# FIRRTL Language Definition
+
+``` ebnf
+(* Whitespace definitions *)
+indent = " " , { " " } ;
+dedent = ? remove one level of indentation ? ;
+newline = ? a newline character ? ;
+
+(* Integer literal definitions  *)
+digit_bin = "0" | "1" ;
+digit_oct = digit_bin | "2" | "3" | "4" | "5" | "6" | "7" ;
+digit_dec = digit_oct | "8" | "9" ;
+digit_hex = digit_dec
+          | "A" | "B" | "C" | "D" | "E" | "F"
+          | "a" | "b" | "c" | "d" | "e" | "f" ;
+int = '"' , "b" , [ "-" ] , { digit_bin } , '"'
+    | '"' , "o" , [ "-" ] , { digit_oct } , '"'
+    | '"' , "h" , [ "-" ] , { digit_hex } , '"'
+    |             [ "-" ] , { digit_bin } ;
```

Comment:
```suggestion
posint = '"' , "b" , { digit_bin } , '"'
       | '"' , "o" , { digit_oct } , '"'
       | '"' , "h" , { digit_hex } , '"'
       |              { digit_bin } ;
negint = '"' , "b" , [ "-" ] , { digit_bin } , '"'
       | '"' , "o" , [ "-" ] , { digit_oct } , '"'
       | '"' , "h" , [ "-" ] , { digit_hex } , '"'
       |             [ "-" ] , { digit_bin } ;
int = (posint | negint)
```

Reply:
Definitely a nit and does not really add anything if spec parseability is not a concern

Reply:
I'll break this up into `nat` (natural number) separate from `int` in a fixup commit.